### ISSUE TITLE
Deprecate `user_err` and use instead an error registering mechanism.

### ISSUE THREAD

**Issue by ejgallego** [coq/coq#7560]
In the current codebase, errors are generated in a totally ad-hoc way using the generic `user_err` function plus an arbitrary message.

This approach has many problems, including #7373. Indeed, as a first step to solve #7373 I propose we deprecate the `user_err` message and instead we require error producers to register errors in the same way it is done for warnings. Thus, we should:

1. extend `CErrors` so that errors can be registered in the same way than warnings.
2. replace all the uses of `user_err` by the proper, registered error.

Once this is done more effort will be needed to fix #7373 adequately, but at least `CError` could provide now with the list of registered errors in a programatic way.

Another largely orthogonal issue is the implementation of the registration functionality itself. I'd like to use extensible datatypes for this but that would require a newer OCaml (cc: #7522 ), so let's see how that goes, indeed that new implementation shouldn't be a blocker for this feature.

Of course, the implementation of the registration functionality for `CWarnings`, `CErrors` should be shared.

**Comment by SkySkimmer** [comment]
That's a lot of labels. I don't think it fits `kind: infrastructure` though.

**Comment by ejgallego** [comment]
I think it does as we need some build infrastructure / scripts so errors a written down in a format readable by the doc build system and IDEs.

**Comment by maximedenes** [comment]
I'm removing the label because I don't think this fits really what we expect for a "good first issue", but I certainly don't mean that this is not interesting. In fact, I may try to contribute to it.

**Comment by ejgallego** [comment]
> That's a lot of labels. I don't think it fits `kind: infrastructure` though.

You need some infrastructure to actually extract the error list from OCaml code. A ppx to define errors / warnings would work great [and still could allow for compilation without it]

> I don't think this fits really what we expect for a "good first issue"

The idea of that label was that once the infrastructure is in place, we will have [hundreths?] of errors to be classified and ported to it, so this could indeed be a nice task for a beginner.

**Comment by ejgallego** [comment]
I'm updating with some discussion from https://github.com/coq/coq/pull/12347

In particular, such a registration API could make use of the features in 
https://caml.inria.fr/pub/docs/manual-ocaml/extensiblevariants.html

In particular, the example about private extensible variants would be really close to what we want, so errors have to be properly registered with a printer, extractable doc, etc...

**Comment by Zimmi48** [comment]
Commenting to note that in a Zulip discussion @gares pointed out that `CError.user_err` raises an error that is catchable by the tactic monad, and we should ensure when moving to a newer error mechanism to not break this (errors that were caught should continue to be caught, and in fact catchable errors should be the default behavior).

**Comment by ejgallego** [comment]
> Commenting to note that in a Zulip discussion @gares pointed out that `CError.user_err` raises an error that is catchable by the tactic monad, and we should ensure when moving to a newer error mechanism to not break this property (errors that were caught should continue to be caught, and in fact catchable errors should be the default behavior).

Yup, that should be fine. Adding error registration opens the door to some interesting features, for example we could have a small language for specifying fixes and suggestions, and provide errors with a bit more structure that is usually much welcome by users.

**Comment by fakusb** [comment]
> Adding error registration opens the door to some interesting features, for example we could have a small language for specifying fixes and suggestions, and provide errors with a bit more structure that is usually much welcome by users.

Also, the eventually arriving language server could suggest such fixes to the IDE. The same holds for warnings, for example "Proof using" suggestions or the like.
